_source/_posts/2021-09-12-spring-boot-spa.md

```diff
@@ -0,0 +1,686 @@
+This example resource server defines one endpoint, `/howcaffeinatedami`, that returns a random (and hopefully humorous) string designating your developer's current caffeine level. 
+
+There is also an inner class (at the bottom) that extends `WebSecurityConfigurerAdapter` to configure Spring Security for the application. The configuration class does three things: 
+
+1. ensures that all requests require authentication; 
+
+2. configures Spring Boot to use the standard resource server security configuration with JSON Web Tokens (JWT); and 
+
+3. enables CORS (Cross-Origin Resource Sharing).
+
+You also need to add a line to the `application.properties` file. This line changes the resource server port from the default `8080` to `8082`. You have to do this because the VueJS application will run on `8080`.
+
+`src/main/resources/application.properties`
+
+```prop
+server.port=8082
+```
+
+You now have the code in place for the resource server. You'll still need to configure the OAuth 2.0 and OIDC settings to use Okta as the provider, but first, I want to explain a little about CORS (Cross-Origin Resource Sharing). It's not just an annoying error your browsers throw at you when you're trying to develop apps locally! If you already totally understand CORS or are impatient to keep moving, feel free to skip the next section.
+
+## Spring Boot and CORS
+
+CORS is a protocol that allows for browsers and servers to explicitly allow cross-origin resource sharing. By default, Javascript running in a browser is restrained by a **same-origin policy**. This means that a Javascript application can only make calls to a server that resides on the same domain from which it was loaded. As a starting point, this makes a lot of sense from a security perspective and prevents a lot of abuse. However, this policy would be pretty restrictive if developers were not able to allow exceptions.
+
+CORS allows resource servers to explicitly enable cross-origin requests, to tell the browser what type of requests it allows and from what origins. It is a white-listing scheme in the sense that all cross-origin requests will be denied by the browser unless the server is explicitly configured to allow them. 
+
+Notice that this protocol is mediated between the server and the browser. The client application, the Vue app in this case, doesn't really have to do anything about CORS. From the perspective of the client app, the HTTP request to the resource server will either succeed or fail. If it fails because of CORS, the request will return a `401 (Unauthorized)`.
+
+In this example, the resource server resides on `http://localhost:8082` and the Vue client will be loaded from from the Vue development server at `http://localhost:8080`. Because the port number is part of the domain, these two URLs are considered different origins by the browser. When the Vue app tries to make a request to the Spring Boot resource server, the browser will throw an error unless CORS is properly configured.
```

Comment:
```suggestion
In this example, the resource server resides on `http://localhost:8082` and the Vue client will be loaded from the Vue development server at `http://localhost:8080`. Because the port number is part of the domain, these two URLs are considered different origins by the browser. When the Vue app tries to make a request to the Spring Boot resource server, the browser will throw an error unless CORS is properly configured.
```
removed redundant word